README.md
# PyAS2

Pyas2 is an AS2 server/client written in python and built on the django framework. The application supports AS2 version 1.2 as defined in the RFC 4130.

# Requirements

* Python (2.6.5+, 2.7)
* Django (1.7, 1.8)
* m2crypto (This is dependent on openssl, which will need to be insatalled seperately in case it is absent.)
* requests
* pyasn1
* cherrypy (Optional if you want to run server using management command runas2server)
* pyinotify on *nix (Optional for using the send daemon)
* Python for Windows extensions (pywin) for windows (Optional for using the send daemon)

# Installation

Install using `pip`...

    pip install pyas2

Create a new `django` project

    django-admin.py startproject django_pyas2

Add `'pyas2'` to your `INSTALLED_APPS` setting.

    INSTALLED_APPS = (
        ...
        'pyas2',
    )
Ensure that pyas2 is placed above 'django.contrib.admin' as we are overriding the admin templates.

Include the pyAS2 URLconf in your project urls.py like this::

  `url(r'^pyas2/', include('pyas2.urls')),`
  
Run `python manage.py migrate` to create the pyas2 models and `python manage.py createsuperuser` to create an admin user.

We can also configure a couple of settings for our API, this is not mandartory.
Add the following to your `settings.py` module:
```python
PYAS2 = {
    'ENVIRONMENT' : 'production', ## as2 server in development or production. default is production
    'PORT' : 8888, ## port the as2 server listens. default is 8080
    ## Specift in order to enable ssl/https for the as2 server:
    'SSLCERTIFICATE' : '/path_to_cert/server_cert.pem', 
    'SSLPRIVATEKEY' : '/path_to_cert/server_privkey.pem',
    'ROOTDIR' : '/path_to_datadir/data', ## Full path to the base directory for storing files, logs ... 
    'PYTHONPATH' : '/path_to_python/python', ## Path to the python executable, neccessary with virtual environments
    'ENVIRONMENTTEXT' : 'BETA',  #environment_text: text displayed on right of the logo. Useful to indicate different environments.
    'ENVIRONMENTTEXTCOLOR' : 'Yellow', ## environment_text_color: Use HTML valid "color name" or #RGB values. Default: Black (#000000)
    'LOGLEVEL' : 'DEBUG', ## level for logging to log file. Values: DEBUG,INFO,STARTINFO,WARNING,ERROR or CRITICAL. Default: INFO
    'LOGCONSOLE' : True, ## console logging on (True) or off (False); default is True.
    'LOGCONSOLELEVEL' : 'DEBUG', # level for logging to console/screen. Values: DEBUG,INFO,STARTINFO,WARNING,ERROR or CRITICAL. Default: STARTINFO  
    'MAXRETRIES': 5,    # Maximum number of retries for failed outgoing messages, defaule is 10
    'MDNURL' : 'https://192.168.1.115:8888/pyas2/as2receive', # Return url for receiving async MDNs from partners
    'ASYNCMDNWAIT' : 30, # Maximum wait time in minutes for asyn MDNs from partner, post which message will be marked as failed
    'MAXARCHDAYS' : 30, # number of days files and messages are kept in storage; default is 30
}
```
## Usage

Now that we have completed installation and configuration of pyAS2 we are ready to start tranferring files.

pyAS2 provides a set of management commands:

To start the AS2 server, run the command `python manage.py runas2server`.
The server frontend will now be accessible at http://127.0.0.1:8080/pyas2/ and can be used to monitor messages and create partners, organizations ...

Messages from your partners can be receieved at http://127.0.0.1:8080/pyas2/as2receive, so this is the link they need to post to for sending messages to you.

To start the AS2 send daemon, run the command `python manage.py runas2daemon`.
This daemon process monitors the outbox folder for each partner and when file is available then it triggers a file transfer to the partner.

To send pending asynchronous MDNs requested by your partners, run the command `python manage.py sendasyncmdn`.

To retry failed as2 communications, run the command `python manage.py retryfailedas2comms`.
Please note that this only retires those outbound messages that failed due to HTTP transmission issues.

## Contributing

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D

## License

GNU GENERAL PUBLIC LICENSE
                       Version 2, June 1991

 Copyright (C) 1989, 1991 Free Software Foundation, Inc., <http://fsf.org/>
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 Everyone is permitted to copy and distribute verbatim copies
 of this license document, but changing it is not allowed.


CHANGELOG.rst
Release History
===============

0.4.6 - 2019-04-17
~~~~~~~~~~~~~~~

* Add support for sending CSV files
* Handle cases where compression can happen before signing

0.4.5 - 2019-03-20
~~~~~~~~~~~~~~~

* Use uuid for filenames and not message id
* More verbose logging of exceptions

0.4.4 - 2019-03-15
~~~~~~~~~~~~~~~

* Increase the length of string fields

0.4.3 - 2018-06-09
~~~~~~~~~~~~~~~

* Send encrypted content as binary data instead of base64

0.4.2 - 2018-06-07
~~~~~~~~~~~~~~~

* Also look for `application/x-pkcs7-signature` when verifying MDN signatures
* Limit size of exception logged to database
* Handle case where Sync MDN does not have a Message ID

0.4.1 - 2018-05-07
~~~~~~~~~~~~~~~

* Also look for `application/x-pkcs7-signature` when verifying signatures

0.4.0 - 2018-01-27
~~~~~~~~~~~~~~~

* Cleaner handling of signature verifications
* Added test cases for sterling b2b integrator message and mdn
* Set `max_length` for file fields to manage long folder names.

0.3.8 - 2018-01-09
~~~~~~~~~~~~~~~~~~

* Give option to download certs from the admin.


0.3.7 - 2018-01-09
~~~~~~~~~~~~~~~~~~

* Use a function to get the certificate upload_to.

0.3.6 - 2018-01-05
~~~~~~~~~~~~~~~~~~

* Added view for downloading certificates from the admin.

0.3.5 - 2017-12-20
~~~~~~~~~~~~~~~~~~

* Renewed the certificates used in the django tests.

0.3.4 - 2017-08-17
~~~~~~~~~~~~~~~~~~

* Add migration to the distribution.

0.3.3 - 2017-04-04
~~~~~~~~~~~~~~~~~~

* Use pagination when listing messages in the GUI, also do not use Datatables.
* Set the request MDN field default value to False.

0.3.2 - 2017-03-07
~~~~~~~~~~~~~~~~~~

* Freeze versions of django and CherryPy in setup.py.

0.3.1 - 2016-10-03
~~~~~~~~~~~~~~~~~~

* Fixed pagination issue where it was showing only 25 messages and mdns.
* Added the admin command cleanas2server for deleting old data and logs.

0.3.0 - 2016-06-28
~~~~~~~~~~~~~~~~~~

* Added django test cases for testing each of the permutations as defined in RFC 4130 Section 2.4.2
* Code now follows the pep-8 standard
* Django admin commands now use argparse instead or optparse

0.2.3 - 2016-04-20
~~~~~~~~~~~~~~~~~~

* Added functionality to customize MDN messages at organization and partner levels.

0.2.2 - 2015-10-12
~~~~~~~~~~~~~~~~~~

* Fixes to take care of changes in Django 1.9.x

0.2.1 - 2015-10-12
~~~~~~~~~~~~~~~~~~

* Updated installation and upgrade documentation.

0.2 - 2015-10-11
~~~~~~~~~~~~~~~~

* Added option to disable verification of public certificates at the time of signature verification.
* Fixed bug in the send daemon.
* Added debug log statements.
* Added some internationlization to model fields.

0.1.2 - 2015-09-07
~~~~~~~~~~~~~~~~~~

* Created readthedocs documentation.
* Fixed bug where inbox and outbox folders were not created on saving partners and orgs.
* Fixed bug where MDN search was failing due to orphaned MDNs.

0.1.1 - 2015-09-04
~~~~~~~~~~~~~~~~~~

* Increased the max length of MODE_CHOICES model field.
* Detect Signature Algorithm from the MIME message for outbound messages.

0.1 - 2015-04-29
~~~~~~~~~~~~~~~~

* Initial release.

.. _`master`: https://github.com/abhishek-ram/pyas2 


init.py
import logging
import os
import sys
from django.conf import settings
from pyas2 import as2utils

### Declare global variables
gsettings = {}
logger = None
convertini2logger = {'DEBUG':logging.DEBUG,'INFO':logging.INFO,'WARNING':logging.WARNING,'ERROR':logging.ERROR,'CRITICAL':logging.CRITICAL,'STARTINFO':25}

## initialize global settings

def initialize():
    global gsettings
    pyas2_settings = {}
    if hasattr(settings, 'PYAS2'):
        pyas2_settings = settings.PYAS2
    if not gsettings:
        gsettings['environment'] = pyas2_settings.get('ENVIRONMENT','production')
        gsettings['port'] = pyas2_settings.get('PORT', 16388)
        gsettings['ssl_certificate'] = pyas2_settings.get('SSLCERTIFICATE',None)
        gsettings['ssl_private_key'] = pyas2_settings.get('SSLPRIVATEKEY',None)
        gsettings['environment_text'] = pyas2_settings.get('ENVIRONMENTTEXT',' ')
        gsettings['environment_text_color'] = pyas2_settings.get('ENVIRONMENTTEXTCOLOR','Black')
        gsettings['root_dir'] = settings.BASE_DIR
        gsettings['python_path'] = pyas2_settings.get('PYTHONPATH', sys.executable)
        gsettings['managepy_path'] = as2utils.join(settings.BASE_DIR, 'manage.py')
        gsettings['daemon_port'] = pyas2_settings.get('DAEMONPORT', 16388)
        if pyas2_settings.get('ROOTDIR') and os.path.isdir(pyas2_settings.get('ROOTDIR')): 
            gsettings['root_dir'] = pyas2_settings.get('ROOTDIR')
        gsettings['payload_receive_store'] = as2utils.join(gsettings['root_dir'], 'messages', '__store', 'payload', 'received')
        gsettings['payload_send_store'] = as2utils.join(gsettings['root_dir'], 'messages', '__store', 'payload', 'sent')
        gsettings['mdn_receive_store'] = as2utils.join(gsettings['root_dir'], 'messages', '__store', 'mdn', 'received')
        gsettings['mdn_send_store'] = as2utils.join(gsettings['root_dir'], 'messages', '__store', 'mdn', 'sent')
        gsettings['log_dir'] = as2utils.join(gsettings['root_dir'], 'logging')
        for sett in ['payload_receive_store', 'payload_send_store', 'mdn_receive_store', 'mdn_send_store', 'log_dir']:
            as2utils.dirshouldbethere(gsettings[sett])
        gsettings['log_level'] = pyas2_settings.get('LOGLEVEL','INFO')
        gsettings['log_console'] = pyas2_settings.get('LOGCONSOLE',False)
        gsettings['log_console_level'] = pyas2_settings.get('LOGCONSOLELEVEL','STARTINFO')
        gsettings['max_retries'] = pyas2_settings.get('MAXRETRIES',30)
        gsettings['mdn_url'] = pyas2_settings.get('MDNURL','http://localhost:8080/pyas2/as2receive')
        gsettings['async_mdn_wait'] = pyas2_settings.get('ASYNCMDNWAIT',30)
        gsettings['max_arch_days'] = pyas2_settings.get('MAXARCHDAYS',30)
        gsettings['minDate'] = 0 - gsettings['max_arch_days']

def initserverlogging(logname):
    # initialise file logging
    global logger
    logger = logging.getLogger(logname)
    logger.setLevel(convertini2logger[gsettings['log_level']])
    handler = logging.handlers.TimedRotatingFileHandler(os.path.join(gsettings['log_dir'],logname+'.log'),when='midnight',backupCount=10)
    fileformat = logging.Formatter("%(asctime)s %(levelname)-9s: %(message)s",'%Y%m%d %H:%M:%S')
    handler.setFormatter(fileformat)
    logger.addHandler(handler)
    # initialise console/screen logging
    if gsettings['log_console']:
        console = logging.StreamHandler()
        console.setLevel(convertini2logger[gsettings['log_console_level']])
        consoleformat = logging.Formatter("%(asctime)s %(levelname)-9s: %(message)s",'%Y%m%d %H:%M:%S')
        console.setFormatter(consoleformat) # add formatter to console
        logger.addHandler(console)  # add console to logger
    return logger
